File: express/src/worker/tasks/ai.ts
import { Job, Worker, Queue } from "bullmq";
import { Tasks, Queues } from "../../types";
import { config } from "./config";
import { huggingFace } from "../../lib/clients/huggingFace";
import { crudPlace } from "../../models/crud";

// Define Queue
const aiQueue = new Queue(Queues.AI, config);

// Embedding Text
interface PlaceEmbeddingData {
    placeId: string;
}

async function placeEmbeddingTask(job: Job<PlaceEmbeddingData>): Promise<void> {
    const { placeId } = job.data;
    const place = await crudPlace.get(placeId);
    if (!place) {
        console.log(`No place with id ${placeId} found in the database`);
        return;
    }
    const text = `${place.title} - ${place.description}`;
    const result = await huggingFace.embedText(text);

    console.log("Finished embedding");
    console.log(result);
}

export const placeEmbedding = (placeId: string) => {
    aiQueue.add(Tasks.PLACE_EMBEDDING, { placeId });
};

// Worker
type AiTasksData = PlaceEmbeddingData;

async function aiTasksProcessor(job: Job<AiTasksData>) {
    switch (job.name) {
        case Tasks.PLACE_EMBEDDING:
            await placeEmbeddingTask(job as Job<PlaceEmbeddingData>);
            break;
        default:
            throw new Error(`Unknown job name: ${job.name}`);
    }
}

const aiWorker = new Worker(Queues.AI, aiTasksProcessor, config);
aiWorker.on("failed", (job, err) => {
    console.error(`Job ${job?.id} failed with error ${err.message}`);
});

File: backend-fastapi/worker/tasks/ai.py
import logging
from typing import TypedDict

from beanie import PydanticObjectId
from bson import ObjectId

from config import settings
from lib.clients import hf_client
from worker.tasks.broker import dramatiq_task

from ..setup import Queues, Tasks


class PlaceEmbbeddingData(TypedDict):
    place_id: str | ObjectId


@dramatiq_task(Tasks.PLACE_EMBEDDING, Queues.AI)
async def place_embedding_task(data: PlaceEmbbeddingData):
    # Lazy loading to avoid circular import issues
    from models.crud import crud_place

    place_id = data["place_id"]
    place = await crud_place.get_document(place_id)
    if place is None:
        logging.warning(f"No place with id {place_id} found in the database")
        return

    text = f"{place.title} - {place.description}"
    result = await hf_client.embed_text(text)
    place.embedding = result
    await crud_place.save_document(place)
    logging.info(result)


def place_embeddding(
    place_id: str | ObjectId | PydanticObjectId | None,
) -> None:
    if settings.is_test or place_id is None:
        return
    place_embedding_task.send(PlaceEmbbeddingData(place_id=str(place_id)))
